ee/localserver/krypto-ec-middleware.go

```diff
@@ -459,3 +477,197 @@ func (bhr *bufferedHttpResponse) WriteHeader(code int) {
+		)
+		return
+	}
+	defer e.presenceDetectionLock.Unlock()
+
+	type detectionResult struct {
+		durtionSinceLastDetection time.Duration
+		err                       error
+	}
+
+	detectionDoneChan := make(chan *detectionResult)
+
+	// kick of presence detection
+	gowrapper.Go(ctx, e.slogger, func() {
+		durationSinceLastDetection, err := e.presenceDetector.DetectPresence(reason, detectionIntervalDuration)
+		if err != nil {
+			e.slogger.Log(ctx, slog.LevelInfo,
+				"presence_detection",
+				"reason", reason,
+				"interval", detectionIntervalDuration,
+				"duration_since_last_detection", durationSinceLastDetection,
+				"err", err,
+			)
+		}
+
+		detectionDoneChan <- &detectionResult{
+			durtionSinceLastDetection: durationSinceLastDetection,
+			err:                       err,
+		}
+	})
+
+	callbackTicker := time.NewTicker(e.presenceDetectionCallbackInterval)
+	defer callbackTicker.Stop()
+
+	presenceDetectionStartTime := time.Now()
+
+	var returnedDetectionResult *detectionResult
+
+	headers := map[string][]string{
+		kolideArchHeaderKey: {runtime.GOARCH},
+		kolideOsHeaderKey:   {runtime.GOOS},
+	}
+
+	kolideSessionId, ok := cmdReq.CallbackHeaders[kolideSessionIdHeaderKey]
+	if ok {
+		headers[kolideSessionIdHeaderKey] = kolideSessionId
+	}
+
+	for {
+		var callbackBody map[string]string
+
+		if returnedDetectionResult != nil {
+			// presence detection complete
+			callbackBody = map[string]string{
+				"msg":          "presence detection completed",
+				"elapsed_time": time.Since(presenceDetectionStartTime).String(),
+				kolideDurationSinceLastPresenceDetectionHeaderKey: returnedDetectionResult.durtionSinceLastDetection.String(),
+			}
+
+			if returnedDetectionResult.err != nil {
+				e.slogger.DebugContext(ctx,
+					"presence detection error",
+					"err", returnedDetectionResult.err,
+				)
+				callbackBody["err"] = returnedDetectionResult.err.Error()
+			}
+
+			headers[kolideDurationSinceLastPresenceDetectionHeaderKey] = []string{returnedDetectionResult.durtionSinceLastDetection.String()}
+		} else {
+			// still waiting on presence detection
+			callbackBody = map[string]string{
+				"msg":          "waiting for user to complete presence detection",
+				"elapsed_time": time.Since(presenceDetectionStartTime).String(),
+			}
+		}
+
+		callBackData, err := e.presenceDetectionCallbackKryptoResponse(challengeBox, headers, callbackBody)
+		if err != nil {
+			e.slogger.ErrorContext(ctx,
+				"error creating presence detection callback response",
+				"err", err,
+			)
+			return
+		}
+
+		req, err := cmdReq.CallbackReq()
+		if err != nil {
+			e.slogger.ErrorContext(ctx,
+				"error creating presence detection callback request",
+				"err", err,
+			)
+			return
+		}
+
+		sendCallback(e.slogger, req, callBackData)
+
+		if returnedDetectionResult != nil {
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			e.slogger.InfoContext(ctx,
+				"presence detection timed out",
```

Comment:
Useful to add an `elapsed` to this log too?

Reply:
done